CRM/Lonesome/Form/Search/NoGroup.php

```diff
@@ -136,18 +109,35 @@ function from() {
    * @return string, sql fragment with conditional expressions
    */
   function where($includeContactIDs = FALSE) {
-    $params = array();
+      
+    $withoutGroups = CRM_Utils_Array::value('without_groups',
+      $this->_formValues
+    );
+    $whereWithoutGroups = '';
+    if ($withoutGroups){
+      $whereWithoutGroups = "
+        AND NOT EXISTS (
+            SELECT 1 FROM
+            civicrm_group_contact c2
+            WHERE c2.contact_id = contact_a.id)
+        AND NOT EXISTS ( -- dont forget to check smart groups
+            SELECT 1 FROM
+            civicrm_group_contact_cache c3
+            WHERE c3.contact_id = contact_a.id)
+        ";
+    }
+    
+    $withoutTags = CRM_Utils_Array::value('without_tags',
+      $this->_formValues
+    );
+    $whereWithoutTags = '';
+    if ($withoutTags) {
+      $whereWithoutTags .= " AND NOT EXISTS (SELECT 1 FROM civicrm_entity_tag c4 WHERE c4.entity_table = 'civicrm_contact' AND c4.entity_id = contact_a.id)";
```

Comment:
Not anal about it, but try to avoid super long lines. And btw, congratulations to have found civicrm_entity_tag table+entity_table=civicrm_contact!